pkg/server/server_int_test.go

```diff
@@ -0,0 +1,354 @@
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/errors"
+	"github.com/lib/pq"
+	"github.com/stretchr/testify/require"
+)
+
+func TestStartupFailure(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.IgnoreLint(t, "this test should be used to reproduce failures from random tests")
+
+	// Ranges that would trigger startup failure if retry is disabled/removed:
+	// /System/NodeLivenessMax (keys.BootstrapVersionKey)
+	// /Table/11 (keys.SystemSQLCodec.TablePrefix(11))
+	// can be used for testing purposes.
+
+	// Set faultyKey to suspected range that is causing startup failure, comment
+	// out skip statement above and run the test.
+	faultyKey := keys.BootstrapVersionKey
+	runCircuitBreakerTestForKey(t,
+		func(spans []roachpb.Span) (good, bad []roachpb.Key) {
+			for _, span := range spans {
+				if span.ContainsKey(faultyKey) {
+					bad = append(bad, span.Key)
+				} else {
+					good = append(good, span.Key)
+				}
+			}
+			return good, bad
+		})
+}
+
+// TestStartupFailureRandomRange picks one random range and runs the test on it.
+// If this test failed, then suspicious range start key deduced from error could
+// be used in TestStartupFailure to investigate code path not covered by startup
+// retries.
+func TestStartupFailureRandomRange(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+	skip.UnderStressRace(t, "5 nodes with replication is too slow for stress race")
+
+	rng, seed := randutil.NewTestRand()
+	t.Log("TestStartupFailureRandomRange using seed", seed)
+	runCircuitBreakerTestForKey(t,
+		func(spans []roachpb.Span) (good, bad []roachpb.Key) {
+			badRange := rng.Intn(len(spans))
+			t.Log("triggering loss of quorum on range", spans[badRange].Key.String())
+			bad = append(bad, spans[badRange].Key)
+			good = make([]roachpb.Key, len(spans)-1)
+			i := 0
+			for ; i < badRange; i++ {
+				good[i] = spans[i].Key
+			}
+			i++
+			for ; i < len(spans); i++ {
+				good[i-1] = spans[i].Key
+			}
+			return good, bad
+		})
+}
+
+// runCircuitBreakerTestForKey causes ranges identified by passed selector
+// function to lose quorum on node stop. It then triggers circuit breaker by
+// probing those ranges. After all ranges are probed, it restarts last killed
+// node to restore quorum and checks that startup succeeds.
+// Node lifecycle:
+// 1, 2, 3 - live
+// 4, 5    - stopped
+// 6       - stopped and restarted
+func runCircuitBreakerTestForKey(
+	t *testing.T, faultyRangeSelector func([]roachpb.Span) (good, bad []roachpb.Key),
+) {
+	const (
+		nodes = 6
+	)
+
+	ctx := context.Background()
+
+	lReg := testutils.NewListenerRegistry()
+	defer lReg.Close()
+	reg := server.NewStickyInMemEnginesRegistry()
+	defer reg.CloseAllStickyInMemEngines()
+	args := base.TestClusterArgs{
+		ServerArgsPerNode: make(map[int]base.TestServerArgs),
+		ReusableListeners: true,
+	}
+	var enableFaults atomic.Bool
```

Comment:
is this WIP? This atomic is never queried. Is the idea in these tests to start a node against a cluster where some/all ranges have circuit breakers enabled? That is nice! Let me know if you want to figure out how to do this, I'm unsure of whether these tests are up for review or not.